### PR TITLE
Enforce max number of callbacks per workflow

### PR DESCRIPTION
**Pull request by rodrigozhou**
## What changed?

Enforce max number of callbacks that can be attached to a workflow

## Why?

Prevent too many callbacks from making the mutable state size explode.

## How did you test it?

Added unit tests.

### CODE REVIEW

**Comment by dandavison** [comment]
Hey @rodrigozhou, would you mind filling out the "Why" section with context on why this work is being done now, or what bugs / tickets it is addressing?

File: service/frontend/service.go [hunk 1]
```diff
@@ -299,6 +300,7 @@ func NewConfig(
 		EnableNexusAPIs:             dc.GetBoolProperty(dynamicconfig.FrontendEnableNexusAPIs, false),
 		EnableCallbackAttachment:    dc.GetBoolPropertyFnWithNamespaceFilter(dynamicconfig.FrontendEnableCallbackAttachment, false),
 		CallbackURLMaxLength:        dc.GetIntPropertyFilteredByNamespace(dynamicconfig.FrontendCallbackURLMaxLength, 1000),
+		MaxCallbacksPerWorkflow:     dc.GetIntPropertyFilteredByNamespace(dynamicconfig.FrontendMaxCallbacksPerWorkflow, 32),
```
Comment (bergundy) [inline service/frontend/service.go:303]: Love that you've used a round number here.

File: service/frontend/workflow_handler.go [hunk 2]
```diff
@@ -4334,6 +4334,16 @@ func (wh *WorkflowHandler) validateWorkflowCompletionCallbacks(
 		)
 	}
 
+	if len(callbacks) > wh.config.MaxCallbacksPerWorkflow(ns.String()) {
```
Comment (bergundy) [inline service/frontend/workflow_handler.go:4337]: In the future we'll have alternative ways to attach callbacks to workflows (attach to a running workflow / attach to a workflow update).
Additional validation will be required in mutable state but this is a good enough starting point.

File: tests/callbacks_test.go [hunk 3]
```diff
@@ -88,57 +88,69 @@ func (s *FunctionalSuite) TestWorkflowCallbacks_InvalidArgument() {
...
 	dc := s.testCluster.host.dcClient
 	dc.OverrideValue(s.T(), dynamicconfig.FrontendCallbackURLMaxLength, 50)
+	dc.OverrideValue(s.T(), dynamicconfig.FrontendMaxCallbacksPerWorkflow, 2)
 	defer dc.RemoveOverride(dynamicconfig.FrontendEnableCallbackAttachment)
 	defer dc.RemoveOverride(dynamicconfig.FrontendCallbackURLMaxLength)
+	defer dc.RemoveOverride(dynamicconfig.FrontendMaxCallbacksPerWorkflow)
 
 	for _, tc := range cases {
 		tc := tc
 		// TODO: use sdkClient when callbacks are exposed
```
Comment (bergundy) [inline tests/callbacks_test.go:131]: Could you please remove this TODO? We're not exposing callbacks in the client any time soon.

**Review by bergundy — APPROVED** [review]
LGTM, thanks!

File: service/frontend/service.go (hunk 1 above)
Comment (rodrigozhou) [reply service/frontend/service.go:303]: It's a round number base 2. :P

**Pull request merged**